adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java

```diff
@@ -522,4 +525,51 @@ public Pathway addPathway(final Pathway newPathway) throws IllegalArgumentExcept
     public Pathway getPathwayById(final String pathwayId) {
         return pathwayPerId.get(pathwayId);
     }
+
+    /**
+     * Add a {@link ShapePoint} to a shape. A shape is a list of{@link ShapePoint} whereas a {@link ShapePoint}
+     * represents a row from shapes.txt to this. Return the entity added to the repository if the entity was
+     * successfully added, and returns null if the provided newShapePoint already exists in the repository. This method
+     * adds the {@link ShapePoint} to this {@link GtfsDataRepository} while maintaining the order according to the
+     * value of this {@link ShapePoint} shape_pt_sequence.
+     *
+     * @param newShapePoint the internal representation of a row from shapes.txt to be added to the repository.
+     * @return Return the entity added to the repository if the entity was successfully added, and returns null if the
+     * provided newShapePoint already exists in the repository.  This method adds the {@link ShapePoint} to this
+     * {@link GtfsDataRepository} while maintaining the order according to the value of this {@link ShapePoint}
+     * shape_pt_sequence.
+     * @throws IllegalArgumentException if the shape point passed as argument is null
+     */
+    @Override
+    public ShapePoint addShapePoint(final ShapePoint newShapePoint) throws IllegalArgumentException {
+        if (newShapePoint != null) {
+            final String shapeId = newShapePoint.getShapeId();
+            if (shapePerIdShapePtSequence.containsKey(shapeId)) {
+                if (shapePerIdShapePtSequence.get(shapeId).containsKey(newShapePoint.getShapePtSequence())) {
+                    return null;
+                }
+                shapePerIdShapePtSequence.get(shapeId).put(newShapePoint.getShapePtSequence(), newShapePoint);
+            } else {
+                final Map<Integer, ShapePoint> innerMap = new TreeMap<>();
+                innerMap.put(newShapePoint.getShapePtSequence(), newShapePoint);
+                shapePerIdShapePtSequence.put(shapeId, innerMap);
+            }
+            return newShapePoint;
+        } else {
+            throw new IllegalArgumentException("Cannot add null shape point to data repository");
+        }
+    }
+
+    /**
+     * Return the collection of shape points from shapes.txt related to the id provided as parameter; which represents a
+     * shape object. The returned collection is ordered by shape_pt_sequence.
+     *
+     * @param shapeId the key from shapes.txt related to the Route to be returned
+     * @return  the collection of shape points from shapes.txt related to the id provided as parameter; which represents
+     * a shape object. The returned collection is ordered by shape_pt_sequence.
+     */
+    @Override
+    public Collection<ShapePoint> getShapeById(final String shapeId) {
+        return new ArrayList<>(shapePerIdShapePtSequence.get(shapeId).values());
```

Comment:
Here, I wonder if we should query ShapePoint on their shape_pt_sequence (this would required this method to return a TreeMap), or keep the collection of ShapePoint as a list

Reply:
I think the main use case for this method will be to get a sequence of points (the "shape") given a shape ID, so I don't think we need to query on shape_pt_sequence here? Unless I'm misunderstanding the question.

I think there are two other items to discuss here (and generally for the repo for all entities), though:
1. Should we be returning a List vs Map? The former (as implemented) prevents calling code from doing lookups based on shape_pt_sequence - maybe that's what you're saying above @lionel-nj?
1. Should we be doing a deep copy here (as currently implemented) or can we use `Collections.unmodifiableMap()` to return an unmodifiable view of this (and avoid copying it)?

Reply:
Thanks, yes that answers my question.
We can discuss the points that you mentioned in our next meeting.

Reply:
[PR#88 ](https://github.com/MobilityData/gtfs-validator/pull/88), is linked to this same discussion also.

Reply:
As discussed, we will implement a solution using  `Collections.unmodifiableMap()` to avoid data duplication.